c++/src/kj/common.h

```diff
@@ -1928,6 +1929,51 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     return true;
   }
 
+  inline auto operator<=>(decltype(nullptr)) const { 
+    using ReturnType = decltype(ptr[0] <=> ptr[0]);
+    return size_ == 0 ? (ReturnType) std::strong_ordering::equal
+                      : (ReturnType) std::strong_ordering::greater;
+  }
+
+  inline auto operator<=>(const ArrayPtr& other) const {
+    //Lexographic 3-way comparitor function 
+    //NOTE: Should not be used to erase equality, as does not short circuit on size
+    using ReturnType = decltype(ptr[0] <=> other[0]);
+    size_t comparisonSize = kj::min(size_, other.size_);
+    if constexpr (isIntegral<RemoveConst<T>>()) {
+      int ret = memcmp(ptr, other.ptr, comparisonSize * sizeof(T));
+      if(ret) {
+        return ret < 0 ? (ReturnType) std::strong_ordering::less 
+                        : (ReturnType) std::strong_ordering::greater;
+      }
+    } else {
+      for(size_t i = 0; i < comparisonSize; i++) {
+        auto ret = ptr[i] <=> other[i];
+        if(ret != (ReturnType) std::strong_ordering::equal) {
+          return ret;
+        }
+      }
+    }
+    return size_ == other.size_ ? (ReturnType) std::strong_ordering::equal 
+          : size_ <  other.size_ ? (ReturnType) std::strong_ordering::less 
+          : (ReturnType) std::strong_ordering::greater;
+  }
+
+  template <typename U>
+  inline auto operator<=>(const ArrayPtr<U>& other) const {
```

Comment:
do we actually use cross-type comparisons?

Reply:
The tests certainly do, I'm not sure if we put them to use anywhere I'll look into it and respond here.